sycl/test-e2e/TriStateExpr.py

```diff
@@ -0,0 +1,176 @@
+from lit.BooleanExpression import BooleanExpression
+
+
+class E2EExpr(BooleanExpression):
+    build_specific_features = {
+        "build-and-run-mode",
+        "target-spir",
+        "target-nvidia",
+        "target-amd",
+        "native_cpu",
+        "linux",
+        "system-linux",
+        "windows",
+        "system-windows",
+        "enable-perf-tests",
+        "opencl_icd",
+        "cuda_dev_kit",
+        "zstd",
+        "vulkan",
+        "true",
+        "false",
+    }
+
+    def __init__(self, string, variables, build_only_mode, ignore_value):
+        BooleanExpression.__init__(self, string, variables)
+        self.build_only_mode = build_only_mode
+        self.ignore = False
+        self.ignore_value = ignore_value
+
+    @staticmethod
+    def evaluate(string, variables, build_only_mode, ignore_value=True):
+        try:
+            parser = E2EExpr(string, set(variables), build_only_mode, ignore_value)
+            return parser.parseAll()
+        except ValueError as e:
+            raise ValueError(str(e) + ("\nin expression: %r" % string))
+
+    def parseMATCH(self):
+        token = self.token
+        BooleanExpression.parseMATCH(self)
+        if token not in self.build_specific_features and self.build_only_mode:
+            self.ignore = True
+        else:
+            self.ignore = False
+
+    def parseAND(self):
+        self.parseNOT()
+        while self.accept("&&"):
+            left = self.value
+            left_ignore = self.ignore
+            self.parseNOT()
+            right = self.value
+            right_ignore = self.ignore
+            self.value = left and right
+            # Ignore if both are ignore or if one is true and the other is ignore
```

Comment:
> Ignore if both are ignore

Is this actually sound? Wouldn't `runtime_feat_A && !runtime_feat_A` result in `ignore` as its value, while the original expression is always false? I don't know if this might cause real problems though.